# Salary Prediction Demonstration

This notebook demonstrates how the selected Random Forest model can generate a salary prediction for a new employee record. The notebook loads the saved model, prepares a hypothetical employee's information using the same encoded feature structure used during training, and returns an estimated annual salary in U.S. dollars.

This demonstration represents the prediction process that will later be incorporated into the Career Compass Streamlit web application.

In [4]:
# Import libraries used to prepare new data and load the saved model
import pandas as pd
import joblib

## Load the Saved Random Forest Model

In [5]:
# Load the saved Random Forest model and feature information
model_bundle = joblib.load(
    "../models/random_forest_salary_model.joblib"
)

# Extract the model and expected feature columns
salary_model = model_bundle["model"]
feature_columns = model_bundle["feature_columns"]

print("Saved salary prediction model loaded successfully.")
print("Expected encoded features:", len(feature_columns))

Saved salary prediction model loaded successfully.
Expected encoded features: 164


## Define a New Employee Record

The following example represents a senior-level, full-time Data Scientist who resides in the United States, works for a medium-sized U.S. company, and works fully remotely.

In [6]:
# Create a hypothetical employee record using the original dataset categories
new_employee = pd.DataFrame([
    {
        "work_year": 2022,
        "experience_level": "SE",
        "employment_type": "FT",
        "job_title": "Data Scientist",
        "employee_residence": "US",
        "remote_ratio": 100,
        "company_location": "US",
        "company_size": "M"
    }
])

# Display the new employee information
new_employee

,work_year,experience_level,employment_type,job_title,employee_residence,remote_ratio,company_location,company_size
0,2022,SE,FT,Data Scientist,US,100,US,M


### Employee Category Key

| Variable | Selected Value | Meaning |
|---|---|---|
| Work year | 2022 | Salary prediction year |
| Experience level | SE | Senior-Level |
| Employment type | FT | Full-Time |
| Job title | Data Scientist | Employee's professional role |
| Employee residence | US | United States |
| Remote ratio | 100 | Fully Remote |
| Company location | US | United States |
| Company size | M | Medium-Sized Company |

## Prepare the New Record for Prediction

The saved Random Forest model expects the same encoded feature columns used during training. The new employee record is therefore converted into one-hot encoded columns and aligned with the model's expected feature structure. Any expected columns not represented by the new record are assigned a value of zero.

In [7]:
# One-hot encode the new employee record
#
# drop_first=False is used here because this DataFrame contains only one row.
# The result is then aligned with the columns used during model training.
new_employee_encoded = pd.get_dummies(
    new_employee,
    drop_first=False
)

# Align the new record with the exact training feature structure
new_employee_encoded = new_employee_encoded.reindex(
    columns=feature_columns,
    fill_value=0
)

# Verify the final dimensions
print("New employee feature dimensions:", new_employee_encoded.shape)
print("Expected number of features:", len(feature_columns))

New employee feature dimensions: (1, 164)
Expected number of features: 164


In [8]:
# Display only the active nonzero features for this employee
active_features = new_employee_encoded.loc[
    :,
    new_employee_encoded.iloc[0] != 0
]

active_features

,work_year,remote_ratio,experience_level_SE,employment_type_FT,job_title_Data Scientist,employee_residence_US,company_location_US,company_size_M
0,2022,100,True,True,True,True,True,True


## Generate the Salary Prediction

The prepared employee record is passed to the saved Random Forest model. The model returns an estimated annual salary in U.S. dollars.

In [9]:
# Generate a salary prediction for the hypothetical employee
predicted_salary = salary_model.predict(
    new_employee_encoded
)[0]

# Display the prediction in a readable currency format
print("Career Compass Salary Prediction")
print("-" * 40)
print(f"Estimated Annual Salary: ${predicted_salary:,.2f}")

Career Compass Salary Prediction
----------------------------------------
Estimated Annual Salary: $168,265.00


In [10]:
# Create a clean summary of the employee and model prediction
prediction_summary = pd.DataFrame({
    "Employee Characteristic": [
        "Work Year",
        "Experience Level",
        "Employment Type",
        "Job Title",
        "Employee Residence",
        "Remote Work Ratio",
        "Company Location",
        "Company Size",
        "Predicted Annual Salary"
    ],
    "Value": [
        "2022",
        "Senior-Level",
        "Full-Time",
        "Data Scientist",
        "United States",
        "Fully Remote (100%)",
        "United States",
        "Medium",
        f"${predicted_salary:,.2f}"
    ]
})

prediction_summary

,Employee Characteristic,Value
0,Work Year,2022
1,Experience Level,Senior-Level
2,Employment Type,Full-Time
3,Job Title,Data Scientist
4,Employee Residence,United States
5,Remote Work Ratio,Fully Remote (100%)
6,Company Location,United States
7,Company Size,Medium
8,Predicted Annual Salary,"$168,265.00"


## Prediction Demonstration Summary

The demonstration confirms that the saved Random Forest model can be loaded independently and used to generate salary estimates for new employee records. The hypothetical employee information was transformed into the same encoded feature structure used during model training before being passed to the model.

This process represents the core scoring workflow planned for the Career Compass application. In a future deliverable, the same steps will be incorporated into a Streamlit interface so users can select employee characteristics from a form and receive a salary estimate without interacting directly with Python code.

The prediction should be interpreted as a decision-support estimate rather than an exact compensation recommendation. Actual salaries may also depend on factors that are not represented in the current dataset.

## Prediction Interpretation

The Random Forest model estimates that this hypothetical employee would earn approximately **$168,265 annually** based on the employee characteristics provided. The prediction reflects patterns learned from the historical salary dataset and should be interpreted as an informed estimate rather than an exact salary recommendation.

Actual compensation may differ due to additional factors that are not represented in the current dataset, including education, certifications, years of experience, negotiation, company-specific compensation policies, and regional labor market conditions.

In [11]:
print("=" * 50)
print("      CAREER COMPASS SALARY PREDICTION")
print("=" * 50)

print(f"\nJob Title: {new_employee['job_title'][0]}")
print(f"Experience Level: Senior-Level")
print(f"Employment Type: Full-Time")
print(f"Company Size: Medium")
print(f"Location: United States")
print(f"Remote Work: 100%")

print("\nPredicted Annual Salary")
print("------------------------")
print(f"${predicted_salary:,.2f}")

print("\nPrediction Complete")

      CAREER COMPASS SALARY PREDICTION

Job Title: Data Scientist
Experience Level: Senior-Level
Employment Type: Full-Time
Company Size: Medium
Location: United States
Remote Work: 100%

Predicted Annual Salary
------------------------
$168,265.00

Prediction Complete
